src/content/blog/content-layer-deep-dive.mdx

```diff
@@ -0,0 +1,242 @@
+		const response = await fetch('https://restcountries.com/v3.1/all');
+		const data = await response.json();
+		// Must return an array of entries with an id property, or an object with IDs as keys and entries as values
+		return data.map((country) => ({
+			id: country.cca3,
+			...country,
+		}));
+	},
+	// optionally define a schema using Zod
+	schema: z.object({
+		id: z.string(),
+		name: z.string(),
+		capital: z.array(z.string()),
+		population: z.number(),
+		// ...
+	}),
+});
+
+export const collections = { countries };
+```
+
+This data is then available in your `.astro` components, just like before:
+
+```astro
+---
+// src/pages/countries/[id].astro
+import type { GetStaticPaths } from 'astro';
+import { getCollection } from 'astro:content';
+
+export const getStaticPaths: GetStaticPaths = async () => {
+	const collection = await getCollection('countries');
+	if (!collection) return [];
+	return collection.map((country) => ({
+		params: {
+			id: country.id,
+		},
+		props: {
+			country,
+		},
+	}));
+};
+
+const { country } = Astro.props;
+---
+
+<h1>{craft.data.name}</h1>
+<p>Capital: {craft.data.capital}</p>
+```
+
+We were already pretty pleased with this experience of querying and rendering your data on the page, so we turned our attention to the mechanics underlying the process. Let's take that deep dive into how the Content Layer API organizes, manages, and uses your content!
+
+## Content layer lifecycle
+
+When `astro build` or `astro dev` is run, the loader for each collection is invoked in parallel. These loaders update their own scoped data store, which is preserved between builds.
+
+<BlogContentImage src={astroBuildImage} />
+
+Astro components and pages can then use `getCollection` or `getEntry` to query the data. The data is immutable at that point, so the pages are all querying the same snapshot, compiled at build time. This applies whether the page is being prerendered at build time, or server rendered on-demand. The important thing to note here is that **the data store is only updated at build time: a deployed site cannot change the data store.** If a data source needs to update a collection it must do this by triggering a new build.
+
+During development the data store can be updated on demand, by using the hotkey `s+enter` when running `astro dev`. Integrations can also update the data during dev. This could be used for things such as registering a development refresh endpoint, or opening a socket to a CMS to listen for updates. This is only available during development though: once deployed the data store is immutable.
+
+## How a loader works
+
+A loader interacts with the content layer via a data store object. This is a key-value store that is scoped to an individual collection. A collection can only access its own entries, but has complete control over these. If a loader knows that its data source hasn't changed then it can skip updating entirely, or it can just update the entries which have changed.
```

Comment:
comma nit
```suggestion
A loader interacts with the content layer via a data store object. This is a key-value store that is scoped to an individual collection. A collection can only access its own entries but has complete control over these. If a loader knows that its data source hasn't changed then it can skip updating entirely, or it can just update the entries which have changed.
```